dot/parachain/util/backing_implicit_view.go

```diff
@@ -0,0 +1,498 @@
+//
+// IMPORTANT: not all blocks are guaranteed to be allowed for some leaves, it may
+// happen that a block info is only kept in the view storage because of a retaining rule.
+//
+// For getting relay-parents that are valid for parachain candidates use `KnownAllowedRelayParentsUnder` method.
+func (view *BackingImplicitView) AllAllowedRelayParents() []common.Hash {
+	allRelayParents := make([]common.Hash, 0, len(view.blockInfoStorage))
+	for relayParent := range view.blockInfoStorage {
+		allRelayParents = append(allRelayParents, relayParent)
+	}
+	return allRelayParents
+}
+
+// ActivateLeaf activates a leaf in the implicit view.
+// This will request the minimum relay parents for the leaf and will load headers in the
+// ancestry of the leaf as needed. These are the 'implicit ancestors' of the leaf.
+//
+// To maximise reuse of outdated leaves, it's best to activate new leaves before
+// deactivating old ones.
+func (view *BackingImplicitView) ActivateLeaf(leafHash common.Hash, subsystemToOverseer chan<- any) error {
+	if _, exists := view.leaves[leafHash]; exists {
+		return fmt.Errorf("%w: %s", errLeafAlreadyKnown, leafHash)
+	}
+
+	fetched, err := view.fetchFreshLeafAndInsertAncestry(leafHash, subsystemToOverseer)
+	if err != nil {
+		return err
+	}
+
+	// Retain at least `minimumRetainLength` blocks in storage.
+	// This helps to avoid Chain API calls when activating leaves in the same chain.
+	retainMinimum := min(
+		fetched.minimumAncestorNumber,
+		max(fetched.leafNumber-minimumRetainLength, 0),
+	)
+	// store the leaf in the active leaves map
+	view.leaves[leafHash] = activeLeafPruningInfo{retainMinimum: retainMinimum}
+	return nil
+}
+
+// Deactivate a leaf in the view. This prunes any outdated implicit ancestors as well.
+// Returns hashes of blocks pruned from storage.
+func (view *BackingImplicitView) DeactivateLeaf(leafHash common.Hash) []common.Hash {
+	// Check if the leaf exists in the active leaves map
+	if _, exists := view.leaves[leafHash]; !exists {
+		return nil
+	}
+
+	// Remove the leaf from the active leaves map
+	delete(view.leaves, leafHash)
+
+	// Determine the minimum retain block number across all remaining leaves
+	minimumBlockNumber := ^parachaintypes.BlockNumber(0) // Max value for BlockNumber
+	for _, leafInfo := range view.leaves {
+		if leafInfo.retainMinimum < minimumBlockNumber {
+			minimumBlockNumber = leafInfo.retainMinimum
+		}
+	}
+
+	// Prune everything before the minimum out of all leaves,
+	// pruning absolutely everything if there are no leaves (empty view)
+
+	// Collect and prune blocks with blockNumber < minimum
+	var leafToRemove []common.Hash
+	for hash, info := range view.blockInfoStorage {
+		if info.blockNumber < minimumBlockNumber {
+			leafToRemove = append(leafToRemove, hash)
+		}
+	}
+
+	for _, hash := range leafToRemove {
+		delete(view.blockInfoStorage, hash)
+	}
+
+	return leafToRemove
```

Comment:
It's safe to delete from maps during iteration.

```suggestion
	var pruned []common.Hash
	for hash, info := range view.blockInfoStorage {
		if info.blockNumber < minimumBlockNumber {
			delete(view.blockInfoStorage, hash)
			pruned = append(pruned, hash)
		}
	}

	return pruned
```